File: agents/Group40/MCTSAgent.py
import numpy as np
import math
from timeit import default_timer as timer
from random import shuffle
# good article to read: https://medium.com/@quasimik/monte-carlo-tree-search-applied-to-letterpress-34f41c86e238
class MCTSNode():
    def __init__(self, colour, move = None):
        self.children = []
        # N = number of simulations that was played involving this node
        self.N = 0
        # Q = Sum of all backpropogated payouts
        self.Q = 0
        # Last move that was made on parent board to get to this node
        # Used to reconstruct the board without saving the entire board in the nodes
        self.move = move
        # Whose turn it is on that node
        self.colour = colour

    def is_leaf(self):
        return len(self.children) == 0
    
class UCT():
    def __init__(self):
        self.C = math.sqrt(2)

    def calculate_value(self, node: MCTSNode, parent: MCTSNode):
        return (node.Q / node.N) + (self.C * math.sqrt(math.log(parent.N) / node.N))

class ExpandAll():
    def __init__(self):
        pass
    
    @staticmethod
    def generate_new_nodes(board, colour):
        moves = get_legal_moves(board)
        nodes = [None] * len(moves)
        for i, move in enumerate(moves):
            node = MCTSNode(colour, move)
            nodes[i] = node
        return nodes
    
def get_opp_colour(colour):
        """Returns the char representation of the colour opposite to the
        current one.
        """
        if colour == "R":
            return "B"
        else: 
            return "R"
        
def get_legal_moves(board: np.ndarray):
    moves = []
    for rowIndex in range(len(board)):
        curRow = board[rowIndex]
        for colIndex in range(len(curRow)):
            cell_color = curRow[colIndex]
            if cell_color == '0':
                moves.append((rowIndex, colIndex))
    return moves

def convert_board_to_string(board):
    stringBoard = ""
    for rowIndex in range(len(board)):
        separator = ","
        if rowIndex == len(board) - 1:
            separator = ""
        stringBoard += "".join(board[rowIndex]) + separator
    return stringBoard
        

def get_winner(board):
    # Use the engine's code to get the winner of a board.
    from Board import Board as engineBoard
    from Colour import Colour 
    ref_board = engineBoard.from_string(convert_board_to_string(board))
    ref_board.has_ended()
    return Colour.get_char(ref_board.get_winner())

class RandomPlayout():
    def __init__(self):
        pass
    
    @staticmethod
    def playout(board, colour):
        # Optimised random playout: get legal moves, shuffle that array and play in that order
        # Then check who won, no need to check after each move
        moves = get_legal_moves(board)
        # Shuffle in place
        shuffle(moves)
        cur_colour = colour
        current_board = np.copy(board)
        for i, move in enumerate(moves):
            current_board[move[0]][move[1]] = cur_colour
            cur_colour = get_opp_colour(cur_colour)
        return get_winner(current_board)


class MCTSBackPropogate():
    def __init__(self):
        pass
    
    @staticmethod
    def update(node: MCTSNode, rWins, bWins):
        totalSimulaitons = rWins + bWins
        node.N += totalSimulaitons
        # Update using opposite colour. On blue nodes, add red wins. On red nodes, add blue wins
        # Because on blue colored nodes, the children are red, so we are choosing the move that results in red node that produces the best outcome for blue
        # So we add blue win on red nodes
        if node.colour == "R":
            node.Q += bWins
        else:
            node.Q += rWins

class RobustChild():
    def __init__(self):
        pass

    @staticmethod
    def get_best_child(root: MCTSNode):
        children = root.children
        bestChild = None
        bestValue = float("-inf")
        for node in children:
            value = node.N
            if value > bestValue:
                bestChild = node
                bestValue = value
        return bestChild

class MCTSAgent():
    def __init__(self):
        # Parameters
        self.simulations_count = 10
        self.time_limit_seconds = 5
        # Policies
        self.selection_policy = UCT()
        self.expansion_policy = ExpandAll()
        self.simulation_policy = RandomPlayout()
        self.back_propogation_policy = MCTSBackPropogate()
        self.root_move_selection_policy = RobustChild()
        
    def expand(self, root: MCTSNode, board: np.ndarray):
       newNodes = self.expansion_policy.generate_new_nodes(board, get_opp_colour(root.colour))
       # Link the newly generated children with the parent
       root.children += newNodes
       return newNodes
    
    def select_best_move(self, root):
        return self.root_move_selection_policy.get_best_child(root).move

    def simulate(self, node, board):
        rWins = 0
        bWins = 0
        for i in range(self.simulations_count):
            winner = self.simulation_policy.playout(board, get_opp_colour(node.colour))
            if winner == "R":
                rWins += 1
            else:
                bWins += 1
        return (rWins, bWins)

    def select(self, root: MCTSNode) -> MCTSNode:
        bestValue = -float("inf")
        bestChild = None
        for child in root.children:
            value = self.selection_policy.calculate_value(child, root)
            if value > bestValue:
                bestValue = value
                bestChild = child
        return bestChild
        
    def MCTS(self, initial_board: np.ndarray, colour: str, turn_count: int):
        self.root = MCTSNode(colour)
        self.colour = colour
        self.opp_colour = get_opp_colour(colour)
        start_time = timer()
        # While not reached time limit, run MCTS 
        while (timer()  - start_time) < self.time_limit_seconds:
            node = self.root
            current_board = np.copy(initial_board)
            # Selection phase
            path = [node]
            while not node.is_leaf():
                node = self.select(node)
                path.append(node)
                move = node.move
                current_board[move[0]][move[1]] = get_opp_colour(node.colour)
            # Expansion phase
            newNodes = self.expand(node, current_board)
            # Simulation phase
            for expandedNode in newNodes:
                simulation_board = np.copy(current_board)
                simulation_board[expandedNode.move[0]][expandedNode.move[1]] = get_opp_colour(expandedNode.colour)
                (rWins, bWins) = self.simulate(expandedNode, simulation_board)
                # Back-propogation phase
                # Update the newly expanded node first
                self.back_propogation_policy.update(expandedNode, rWins, bWins)
                # Update all nodes on path, going from latest node (LIFO)
                for i in range(len(path) - 1, -1, -1):
                    nodeOnPath = path[i]
                    self.back_propogation_policy.update(nodeOnPath, rWins, bWins)
        # Select final move
        return self.select_best_move(self.root)


if __name__ == "__main__":
    agent = MCTSAgent()
    board_size = 11
    initial_board = np.full((board_size, board_size), '0')
    print(agent.MCTS(initial_board, "R", 0))

File: agents/Group40/java/MCTSAgent.java
package javaV;

import javaV.common.Common;
import javaV.common.Move;
import javaV.common.SimulationResult;
import javaV.common.SimulationWins;
import javaV.policies.backPropogation.MCTSBackPropogation;
import javaV.policies.backPropogation.RAVEBackPropogation;
import javaV.policies.expansion.ExpandAll;
import javaV.policies.expansion.ExpandOneRandom;
import javaV.policies.rootMoveSelection.RobustChild;
import javaV.policies.rootMoveSelection.SecureChild;
import javaV.policies.selection.RAVE;
import javaV.policies.selection.UCT;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Stack;

public class MCTSAgent {
    // Parameters
    public static int simulationsCntPerCore = 115;
    public static double timeLimitSeconds = 6.5;
    public static double C = 0.0000001;
    // Policies
    private static final RAVE selectionPolicy = new RAVE();
    private static final ExpandOneRandom expansionPolicy = new ExpandOneRandom();
    private static final RAVEBackPropogation backPropogationPolicy = new RAVEBackPropogation();
    private static final RobustChild rootMoveSelectionPolicy = new RobustChild();
    public static int cores = Runtime.getRuntime().availableProcessors();

    private MCTSNode root;

    private MCTSNode select(MCTSNode root){
        double bestValue = Double.NEGATIVE_INFINITY;
        MCTSNode bestChild = null;
        for (MCTSNode child: root.children){
            final double curValue = RAVE.calculateValue(child, root);
            if (curValue > bestValue){
                bestValue = curValue;
                bestChild = child;
            }
        }
        return bestChild;
    }

    private static MCTSNode[] expand(MCTSNode node, char[][] board){
        MCTSNode[] newNodes = ExpandOneRandom.generateNewNodes(node, board, Common.getOppColour(node.colour));
        Collections.addAll(node.children, newNodes);
//        for (MCTSNode child : newNodes){
//
//        }
        return newNodes;
    }

    private static SimulationResult simulate(MCTSNode node, char[][] board){
        SimulationResult simulationResult = new SimulationResult(Common.boardSize);

        Thread[] threads = new Thread[cores];
        SimulationThread[] threadInfos = new SimulationThread[cores];
        for (int i = 0; i < threads.length; i++){
            threadInfos[i] = new SimulationThread(board, node, simulationsCntPerCore);
            threads[i] = new Thread(threadInfos[i]);
            threads[i].start();

        }

        for (Thread thread : threads) {

            try {
                thread.join();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }

        }
        // Combine results from all threads
        for (SimulationThread threadInfo : threadInfos){
            simulationResult.rWins += threadInfo.simulationResult.rWins;
            simulationResult.bWins += threadInfo.simulationResult.bWins;

        }

//        for(int i = 0; i < simulations_count; i++){
//            char winner = simulationPolicy.playout(board, node.colour);
//            if (winner == 'R'){
//                rWins++;
//            }
//            else{
//                bWins++;
//            }
//        }
        return simulationResult;

    }

    private double getTimeLimit(double turn){
        if (turn <= 33) {
            return 7.2;
        } else if (turn <= 50) {
            return 3.3;
        } else {
            return 0.5;
        }
    }

    private static Move selectBestMove(MCTSNode root){
        return RobustChild.getBestChild(root).move;
    }

    public Move MCTS(char[][] board, char colour, int turn_count){
        root = new MCTSNode(colour);
        timeLimitSeconds = getTimeLimit(Math.ceil(turn_count/2));
        final double msTimeLimit = timeLimitSeconds * 1000;
        final long start_time = System.currentTimeMillis();
        while ((System.currentTimeMillis() - start_time) < msTimeLimit){
            MCTSNode node = root;
            char[][] current_board = Common.copy2dArray(board);
            // Selection phase
            ArrayList<MCTSNode> path = new ArrayList<>();
            path.add(node);
            while (!node.children.isEmpty() && node.children.size() == Common.getNumLegalMoves(current_board)){
                node = select(node);
                path.add(node);
                final Move move = node.move;
                current_board[move.y][move.x] = Common.getOppColour(node.colour);
            }
            // Expansion phase
            final MCTSNode[] expandedNodes = expand(node, current_board);
            // Simulation Phase
            for(MCTSNode expandedNode : expandedNodes){
                char[][] simulationBoard = Common.copy2dArray(current_board);
                final Move move = expandedNode.move;
                simulationBoard[move.y][move.x] = Common.getOppColour(expandedNode.colour);
                final SimulationResult simulationResult = simulate(expandedNode, simulationBoard);
                //Back-propogation phase
                //Update the newly expanded node first
                RAVEBackPropogation.update(expandedNode, simulationResult);
                //Update all nodes on path, going from latest node (LIFO)
                for(int i = path.size() - 1; i >= 0; i--){
                    final MCTSNode nodeOnPath = path.get(i);
                    RAVEBackPropogation.update(nodeOnPath, simulationResult);
                }
            }
            // If there are no expanded nodes : reached end, which is unlikely but may happen
            // Simply run simulations on that node
            if (expandedNodes.length == 0){

                final SimulationResult simulationResult = simulate(node, current_board);
                for(int i = path.size() - 1; i >= 0; i--){
                    final MCTSNode nodeOnPath = path.get(i);
                    RAVEBackPropogation.update(nodeOnPath, simulationResult);
                }
            }
        }
        final Move bestMove = selectBestMove(root);
        root = null;
        return bestMove;
    }
}
